lib/src/media/stream/client.dart

```diff
@@ -5,51 +5,112 @@ import 'package:flutter_webrtc/flutter_webrtc.dart';
 
-  StreamClient(this._channel) : _client = StreamServiceClient(_channel) {
+  static final Finalizer<StreamSubscription?> _finalizer = Finalizer((p0) {
+    p0?.cancel();
+  });
+
+  StreamManager(this._channel) : _client = StreamServiceClient(_channel) {
+    _finalizer.attach(this, _errorHandler);
     _channel.rtcPeerConnection.onAddStream = (MediaStream stream) {
-      _stream = stream;
-      _streamController.add(stream);
+      addStream(stream);
     };
 
     _channel.rtcPeerConnection.onConnectionState = (state) {
       if (state == RTCPeerConnectionState.RTCPeerConnectionStateFailed ||
           state == RTCPeerConnectionState.RTCPeerConnectionStateDisconnected) {
         _errorHandler = Stream.periodic(const Duration(seconds: 1)).listen((_) {
-          _streamController.addError(Exception('PeerConnection error'));
+          for (final client in _clients.values) {
+            client._streamController.addError(Exception('PeerConnection error'));
+          }
         });
       }
     };
   }
 
-  String _getValidSDPTrackName(String name) {
+  void addStream(MediaStream stream) {
+    _streams[stream.id] = stream;
+    if (_clients.containsKey(stream.id)) {
+      _clients[stream.id]!._internalStreamController.add(stream);
+    }
+  }
+
+  String getValidSDPTrackName(String name) {
     return name.replaceAll(':', '+');
   }
 
-  Stream<MediaStream> getStream(String name) {
-    if (_stream == null) {
-      final future = add(name);
-      future.onError((error, stackTrace) => _streamController.addError(error ?? Exception('Could not add stream named $name')));
+  StreamClient getStreamClient(String name) {
+    final sanitizedName = getValidSDPTrackName(name);
+    if (_clients.containsKey(sanitizedName)) {
+      return _clients[sanitizedName]!;
+    }
+    final client = StreamClient(name, remove);
+    _clients[sanitizedName] = client;
+
+    if (_streams.containsKey(sanitizedName)) {
+      Future.delayed(const Duration(milliseconds: 100), () {
+        _clients[sanitizedName]!._internalStreamController.add(_streams[sanitizedName]!);
+      });
     } else {
-      _streamController.add(_stream!);
+      final fut = add(sanitizedName);
+      fut.onError((error, stackTrace) => client._streamController.addError(error ?? Exception('Could not add stream named $name')));
     }
-    return _streamController.stream;
+    return client;
   }
 
   Future<void> add(String name) async {
-    await _client.addStream(AddStreamRequest(name: _getValidSDPTrackName(name)));
+    await _client.addStream(AddStreamRequest(name: name));
   }
 
   Future<void> remove(String name) async {
-    await _client.removeStream(RemoveStreamRequest(name: _getValidSDPTrackName(name)));
-    _stream = null;
-    await _errorHandler?.cancel();
+    final sanitizedName = getValidSDPTrackName(name);
+    if (_streams.containsKey(sanitizedName)) {
+      final stream = _streams.remove(sanitizedName)!;
+      await stream.dispose();
```

Comment:
what does `dispose()` do here? under the hood it looks like it just returns `Future.value()` but maybe im looking in the wrong spot?

Reply:
My assumption was that it closes the stream but it looks like no... Can remove